hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/index/RunIndexActionExecutor.java

```diff
@@ -0,0 +1,302 @@
+      // get all instants since the plan completed (both from active timeline and archived timeline)
+      // assumption is that all metadata partitions had same instant upto which they were scheduled to be indexed
+      table.getMetaClient().reloadActiveTimeline();
+      String indexUptoInstant = indexPartitionInfos.get(0).getIndexUptoInstant();
+      List<HoodieInstant> instantsToIndex = getRemainingArchivedAndActiveInstantsSince(indexUptoInstant, table.getMetaClient());
+      LOG.info("Total remaining instants to index: " + instantsToIndex.size());
+
+      // reconcile with metadata table timeline
+      String metadataBasePath = HoodieTableMetadata.getMetadataTableBasePath(table.getMetaClient().getBasePath());
+      HoodieTableMetaClient metadataMetaClient = HoodieTableMetaClient.builder().setConf(hadoopConf).setBasePath(metadataBasePath).build();
+      Set<String> metadataCompletedTimestamps = getCompletedArchivedAndActiveInstantsAfter(indexUptoInstant, metadataMetaClient).stream()
+          .map(HoodieInstant::getTimestamp).collect(Collectors.toSet());
+
+      // index catchup for all remaining instants with a timeout
+      currentIndexedInstant = indexUptoInstant;
+      ExecutorService executorService = Executors.newFixedThreadPool(MAX_CONCURRENT_INDEXING);
+      Future<?> indexingCatchupTaskFuture = executorService.submit(
+          new IndexingCatchupTask(metadataWriter, instantsToIndex, metadataCompletedTimestamps, table.getMetaClient(), metadataMetaClient));
+      try {
+        LOG.info("Starting index catchup task");
+        indexingCatchupTaskFuture.get(config.getIndexingCheckTimeoutSeconds(), TimeUnit.SECONDS);
+      } catch (Exception e) {
+        indexingCatchupTaskFuture.cancel(true);
+        throw new HoodieIndexException(String.format("Index catchup failed. Current indexed instant = %s. Aborting!", currentIndexedInstant), e);
+      } finally {
+        executorService.shutdownNow();
+      }
+      // save index commit metadata and update table config
+      List<HoodieIndexPartitionInfo> finalIndexPartitionInfos = indexPartitionInfos.stream()
+          .map(info -> new HoodieIndexPartitionInfo(
+              info.getVersion(),
+              info.getMetadataPartitionPath(),
+              currentIndexedInstant))
+          .collect(Collectors.toList());
+      HoodieIndexCommitMetadata indexCommitMetadata = HoodieIndexCommitMetadata.newBuilder()
+          .setVersion(LATEST_INDEX_COMMIT_METADATA_VERSION).setIndexPartitionInfos(finalIndexPartitionInfos).build();
+      try {
+        // update the table config and timeline in a lock as there could be another indexer running
+        txnManager.beginTransaction();
+        updateTableConfig(table.getMetaClient(), finalIndexPartitionInfos);
+        table.getActiveTimeline().saveAsComplete(
+            new HoodieInstant(true, INDEX_ACTION, indexInstant.getTimestamp()),
+            TimelineMetadataUtils.serializeIndexCommitMetadata(indexCommitMetadata));
```

Comment:
what happens if we fail before reaching L165. i.e the timeline saving. How do we recover/reconcile